### PR TITLE
Implicit \color, explicitly grouped \textcolor

### PR DESCRIPTION
**Pull request by edemaine**
I've implemented `\color` and `\textcolor` to behave as they do in LaTeX (specifically, the `color` package):

```latex
{\color{red}This is red.}
\textcolor{red}{This is also red}
```

Sadly, this differs from current KaTeX and MathJax which would only make the `T` red in the first example (as if `\color` were actually `\textcolor`).  So this change may break some existing code.  But I think it's for the better, as LaTeX should be the master reference here.

### CODE REVIEW

**Comment by kevinbarabash** [comment]
I didn't realize there was a `color` package.  It's too bad about the difference in behavior.  I think we want this, we just need to communicate the breaking change along with suggested workarounds.

**Comment by kevinbarabash** [comment]
`color` package reference: http://ctan.mirror.rafal.ca/macros/latex/required/graphics/color.pdf

**Comment by edemaine** [comment]
As I understand it, this is the *only* way to change color in LaTeX.  There are other packages, e.g. xcolor, but they build on the same basic interface.

**Comment by xymostech** [comment]
This is admittedly selfish, but KA has a *ton* of math expressions that would break because of this change. Also, our previous `\color` function was compatible with MathJax, which was the reason we implemented it the way we did.

**Comment by edemaine** [comment]
FWIW, MathJax defines a [color extension](http://docs.mathjax.org/en/latest/tex.html#color) that changes the behavior to match LaTeX.  I suppose KaTeX could do the same... though personally I think it makes more sense to match LaTeX and MathJax+color.

**Comment by kevinbarabash** [comment]
@xymostech we're already preprocessing our math so it shouldn't be hard to replace `\color` with `\textcolor`.

**Comment by edemaine** [comment]
@xymostech Another thing we could do is add an `oldColor` or `colorIsTextColor` option to the options set, to enable backwards compatibility mode...

**Comment by gagern** [comment]
I guess you could also declare a macro `\color` which expands to `\textcolor`. That way you don't have to go poking around in some string literal which might one day contain things like `\verb` and `\def` and whatnot, but could still override `\color` to behave like `\textcolor` on a per-size basis.

Nevertheless, this change should come with a bump to 0.8.0, not in the 0.7 line of releases.

File: test/screenshotter/ss_data.yaml [hunk 1]
```diff
@@ -49,9 +49,10 @@ Cases: |
         a-1&\text{otherwise}
     \end{cases}
 Colors:
-    tex: \blue{a}\color{#0f0}{b}\color{red}{c}
+    tex: \blue{a}\textcolor{#0f0}{b}\textcolor{red}{c}
     nolatex: different syntax and different scope
-ColorSpacing: \color{red}{\displaystyle \int x} + 1
+ColorSpacing: \textcolor{red}{\displaystyle \int x} + 1
+ColorImplicit: \color{red}red\color{blue}blue
```
Comment (gagern) [inline test/screenshotter/ss_data.yaml]: I think this test should ensure that a) ending a group restores previous color and b) ending a `\textcolor` does return to the previous implicit color. Something like

```tex
bl{ack\color{red}red\textcolor{green}{green}red\color{blue}blue}black
```

Comment (edemaine) [inline test/screenshotter/ss_data.yaml]: Good point, and nice test!  Changed.

**Comment by edemaine** [comment]
@gagern Good idea.  We could suggest defining a macro `'\color': '\textcolor'` in release notes to 0.8.0, assuming it gets approved at all.  I'll just reiterate that I think it's good to use LaTeX as the reference model.  Even if MathJax is the reference model, it offers both options.

**Comment by gagern** [comment]
Code looks good to me. I'd like a go-ahead from @xymostech, though, to make sure the workaround for the incompatibility is feasible for KA.

**Comment by edemaine** [comment]
@xymostech or @kevinbarabash, could we make a decision here?  I see two reasonable options:

1. Leave `\color` as is (two-argument function) because it's compatible with MathJax without any flags.  But provide an option that switches `\color` to behave as in LaTeX, just like MathJax does.
2. Change `\color` to make it compatible with LaTeX (one-argument function changing current group).  But provide an option that switches `\color` to behave as in MathJax (i.e. old behavior) for (somewhat manual) backwards compatibility.

Option 1 has the advantage of backwards compatibility without option changing.  Option 2 is preferable to me because KaTeX's mantra is/should be "renders just like LaTeX".  But it would probably require a significant version increment.

In both cases, we can provide `\textcolor` which behaves just like LaTeX, because that should unambiguously be a two-argument function.

**Comment by kevinbarabash** [comment]
@edemaine let's go with option 2.  Will want to bump the version number to 0.8.x when we do a publish that includes this change.

**Comment by edemaine** [comment]
Cool, thanks, @kevinbarabash !  That's almost what this PR does.  Preference on calling the option `oldColor` or `colorIsTextColor` or something else?

**Comment by kevinbarabash** [comment]
`colorIsTextColor` is more descriptive.

**Comment by edemaine** [comment]
4dd21a7 adds a `colorIsTextColor`, documentation in README, and `parsesLike` tests that confirm the exact behavior of `\color` in the three cases (no specification, and `colorIsTextColor` set to `false` or `true`).

File: src/functions.js [hunk 2]
```diff
@@ -166,6 +166,14 @@ defineFunction("\\color", {
     };
 });
 
+// \color is handled in Parser.js's parseImplicitGroup
```
Comment (kevinbarabash) [inline src/functions.js]: Thanks for commenting this.  It would be nice to have a way to mark functions in this file as to whether or not the parser needs to the parse an implicit group for them.

File: test/errors-spec.js [hunk 3]
```diff
@@ -224,19 +224,19 @@ describe("Parser.expect calls:", function() {
...
                    "Invalid size: '[' at position 11: \\rule{1em}[̲2em]");
         });
         // Can't test for the [ of an optional group since it's optional
         it("complains about missing } for color", function() {
-            expect("\\color{#ffffff{text}").toFailWithParseError(
-                   "Invalid color: '#ffffff{text' at position 8:" +
-                   " \\color{#̲f̲f̲f̲f̲f̲f̲{̲t̲e̲x̲t̲}");
+            expect("\\textcolor{#ffffff{text}").toFailWithParseError(
+                   "Invalid color: '#ffffff{text' at position 12:" +
+                   " \\textcolor{#̲f̲f̲f̲f̲f̲f̲{̲t̲e̲x̲t̲}");
```
Comment (kevinbarabash) [inline test/errors-spec.js]: I don't even know what that character is, lol.

File: test/katex-spec.js [hunk 4]
```diff
@@ -833,10 +838,26 @@ describe("A color parser", function() {
...
+    it("should use one-argument \\color if requested", function() {
+        expect(oldColorExpression).toParseLike("\\textcolor{#fA6}{xy}", {
+            colorIsTextColor: false,
+        });
+    });
+
+    it("should use two-argument \\color if requested", function() {
+        expect(oldColorExpression).toParseLike("\\textcolor{#fA6}{x}y", {
+            colorIsTextColor: true,
+        });
```
Comment (kevinbarabash) [inline test/katex-spec.js]: Nice tests!

File: test/screenshotter/ss_data.yaml [hunk 5]
```diff
@@ -49,9 +49,10 @@ Cases: |
         a-1&\text{otherwise}
     \end{cases}
 Colors:
-    tex: \blue{a}\color{#0f0}{b}\color{red}{c}
+    tex: \blue{a}\textcolor{#0f0}{b}\textcolor{red}{c}
     nolatex: different syntax and different scope
-ColorSpacing: \color{red}{\displaystyle \int x} + 1
+ColorSpacing: \textcolor{red}{\displaystyle \int x} + 1
+ColorImplicit: bl{ack\color{red}red\textcolor{green}{green}red\color{blue}blue}black
```
Comment (kevinbarabash) [inline test/screenshotter/ss_data.yaml]: nice

**Comment by kevinbarabash** [comment]
@edemaine thanks for the PR.  Sorry for taking so long to make a decision regarding `\color`'s behavior.

File: src/functions.js (hunk 2 above)
Comment (edemaine) [inline src/functions.js]: Yes.  @gagern and I had discussed this at some point, and it felt like *all* the implicit group parsing could be moved to `functions.js`.  Worth considering in the future...

**Comment by edemaine** [comment]
No worries -- big decisions are hard!  Thanks for reviewing + merging.
